CHANGELOG.rst

```diff
@@ -78,15 +79,23 @@ Scripting improvements
 - On MacOS terminals that are not granted permissions to access a folder, ``cd`` would print a spurious "rotten symlink" error, which has been corrected to "permission denied" (:issue:`8264`).
 - Since fish 3.0, for-loops would trigger a variable handler an additional time before the loop was entered. This has been corrected (:issue:`8384`).
 - ``math`` now correctly prints negative values and values larger than ``2**31`` when in hex or octal bases (:issue:`8417`).
+- ``math ncr`` has been sped up (:issue:`8170`).
 - ``dirs`` always produces an exit status of 0, instead of sometimes returning 1 (:issue:`8211`).
 - ``cd ""`` no longer crashes fish (:issue:`8147`).
 - ``set --query`` can now query whether a variable is a path variable via ``--path`` or ``--unpath`` (:issue:`8494`).
 - Tilde characters (``~``) produced by custom completions are no longer escaped when applied to the command line, making it easier to use the output of a recursive ``complete -C`` in completion scripts (:issue:`4570`).
+- ``set --show`` now marks read-only variables (:issue:`8179`).
+- ``LANG=fr_FR.utf8 set --show`` used to crash due to an invalid format string in a translation string, which has been fixed (:issue:`8358`).
+- Builtins ``printf`` and ``math`` used to read the environment variable ``LC_NUMERIC`` to use locale-specific number format. They have been changed to always use ``LC_NUMERIC=C``. In particular ``math``'s decimal separator is now always ".", never "," (:issue:`8204`).
```

Comment:
This is wrong - `math` *always* used `.`. Using `,` doesn't work because that's the argument separator - `ncr(5,2)` is n choose r with n = 5 and r = 2. Even if we worked around that (Excel uses `;` as the argument separator in German!), it would still break expressions brought onto a system with a different locale - this is a common problem with badly ported linux games and shaders! So the only tenable thing to do for any script thing that reads arithmetic expressions is to always use `.` as the decimal separator.

So that is what `math` did from the very beginning. It did that by doing `setlocale(LC_NUMERIC, "C")` while math was being run.

Instead, we now do `setlocale(LC_NUMERIC, "C")` *always*, so for fish internally numbers are always "." and then we override it in the one place where we want localized numbers, which is printf. (tho tbh I would rather we don't do localized numbers there either, but that's a different change)

This makes `math` faster by quite a bit and, on systems supporting `uselocale`, improves thread safety while not affecting printf performance a lot. It also makes it easier for fish to stay correct - 99% of the time we want unlocalized numbers with `.`.

-------

If anything, the user-visible effect is that math is faster.

An additional effect of this was accidentally fixing detection of when we can use posix_spawn on glibc systems with a comma-using locale, which was later properly fixed with https://github.com/fish-shell/fish-shell/commit/f0d4fd85b116a30319c1b42f1f79c33ebaa69441. This increases performance of launching external processes on those systems by ~50% (as in you can run 1.5x the processes in the same time, with things like `command echo`).

Reply:
Wow, thanks for the explanation. I'll drop it since math performance isn't so important. The posix_spawn detection fix is nice though!